### PR TITLE
Incremental merkle tree builder impl (for future tx tree purposes)

### PR DESCRIPTION
**Pull request by jakrawcz-rdx**
Another step towards https://radixdlt.atlassian.net/browse/NODE-537.
As designed in https://radixdlt.atlassian.net/wiki/spaces/RPNV1/pages/2996600937/RNP-7+-+Transaction+Hash+Tree+Inclusion+Exclusion+verification#Incremental-merkle-tree-storage .

The tree struct introduced here is *not yet* used anywhere. Another PR will integrate it into Node's logic.

### CODE REVIEW

File: core-rust/state-manager/src/accumulator_tree/storage.rs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+pub trait AccuTreeStore<N>: ReadableAccuTreeStore<N> + WriteableAccuTreeStore<N> {}
+impl<T: ReadableAccuTreeStore<N> + WriteableAccuTreeStore<N>, N> AccuTreeStore<N> for T {}
+
+/// A vertical slice of an accumulator tree.
+/// This is an "incremental" persistence part of a tree representing a batch of appended leaf nodes
+/// and the resulting merkle updates, propagating up to a single root.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct TreeSlice<N> {
+    /// The tree-levels of this slice, arranged from leafs to root.
+    pub levels: Vec<TreeSliceLevel<N>>,
```
Comment (malexandru-rdx) [inline core-rust/state-manager/src/accumulator_tree/storage.rs:93]: I'm not really convinced by this structure. It's easy to reason about when reading the definition of the struct, but when actually using it seems like a lot of headache. What do you think about a linearized representation of the whole tree?

**Review by malexandru-rdx — COMMENTED** [review]
I am not very convinced with the memory layout.

I think it would shed more light/be closer to the final PR if proof generation would be taken into account from the start. With the current setup seems it will be kind of difficult to do it properly. Here we need to check out if we will only ever need proof for single a transaction or ranges as well (probably we're going to need it when doing ledger sync).

File: core-rust/state-manager/src/accumulator_tree/tree_builder.rs [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+    fn zero() -> Self;
+
+    /// An accumulation function (i.e. giving virtually-unique results for any ordered pair).
+    fn accumulate(left: &Self, right: &Self) -> Self;
+}
+
+/// An accumulator tree with persistence backed by an external storage.
+/// For convenience, this structure tracks the current number of accumulated leaves internally
+/// between calls.
+pub struct AccuTree<'s, S: AccuTreeStore<N>, N> {
```
[the remark this answers is not in the document]
Comment (jakrawcz-rdx) [reply core-rust/state-manager/src/accumulator_tree/tree_builder.rs:82]: I've seen this "accumulator tree" name used in some papers, e.g. https://diem-developers-components.netlify.app/papers/the-diem-blockchain/2020-05-26.pdf and they mean exactly "merkle tree to which leafs are appended":
image

I want to be as far as possible from "hash tree" name, which is used for our "substate store's merkle tree" (of very different nature).

File: core-rust/state-manager/src/accumulator_tree/storage.rs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    pub fn new() -> Self {
+        Self {
+            slices: HashMap::new(),
+        }
+    }
+}
+
+impl<N: Clone> ReadableAccuTreeStore<N> for MemoryAccuTreeStore<N> {
+    fn get_tree_slice(&self, index: usize) -> Option<TreeSlice<N>> {
+        self.slices.get(&index).cloned()
```
[the remark this answers is not in the document]
Comment (jakrawcz-rdx) [reply core-rust/state-manager/src/accumulator_tree/storage.rs:160]: I see...
For "append to tree" and "generate proof" operations, this does not matter (because these operations are `>= O(M + logN)` already, probably with `C` far higher than that of `clone()` :P).
How about we add a dedicated O(1) `get_root_of_slice(index)` when needed?

(alternatively we can return a `&` from trait, but AFAIS our convention seems to be "return actual values from storage")

File: core-rust/state-manager/src/accumulator_tree/tree_builder.rs [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+enum RelativeIndex {
+    CachedLeftSibling,
+    ActualStoredNode(usize),
+    ZeroRightSibling,
+}
+
+struct RelativeNodeAccess<M: Merklizable> {
+    offset: usize,
+    slice_level: TreeSliceLevel<M>,
+    zero: M,
```
[the remark this answers is not in the document]
Comment (jakrawcz-rdx) [reply core-rust/state-manager/src/accumulator_tree/tree_builder.rs:170]: I have this `get()` function, which is supposed to return a reference :/ (the caller only wants reference too)
If I make it return a value (at the cost of `clone()` here and there), then I can totally get rid of `zero` (not needing `PhantomData` either). Is it worth it?

File: core-rust/state-manager/src/accumulator_tree/storage.rs [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+ */
+
+use std::collections::HashMap;
+
+/// The "read" part of an accumulator tree storage SPI.
+pub trait ReadableAccuTreeStore<N> {
+    /// Gets a vertical `TreeSlice` by its end index.
+    /// This index is effectively equal to the number of leaf nodes stored in the tree by _all_ the
+    /// slices _including_ the requested one.
+    fn get_tree_slice(&self, index: usize) -> Option<TreeSlice<N>>;
```
[the remark this answers is not in the document]
Comment (jakrawcz-rdx) [reply core-rust/state-manager/src/accumulator_tree/storage.rs:72]: when we add a `get_slice_which_includes(index)`, we can implement it with db's "get greatest smaller than `index`" operation (which AFAIR, is as performant as "get `index`" in our rocksdb, at least in theory).

File: core-rust/state-manager/src/accumulator_tree/storage.rs (hunk 1 above)
Comment (jakrawcz-rdx) [reply core-rust/state-manager/src/accumulator_tree/storage.rs:93]: (just noting: we are still discussing options with Alexandru here)

In general, this follows the design from https://radixdlt.atlassian.net/wiki/spaces/RPNV1/pages/2996600937/RNP-7+-+Transaction+Hash+Tree+Inclusion+Exclusion+verification#Incremental-merkle-tree-storage which was created with following constraints in mind:

1. There is one tree per epoch header, but we need to persist its state to DB after each ledger header.
   - ideally we should achieve it by writing an amount of data proportional to the number of transactions in that commit (i.e. not proportional to number of transactions in that epoch so far)
2. The storage scheme should be tailored to level-like DBs (i.e. very-cheap inserts; less-cheap updates and deletes; no "append to existing entry" operation).
3. It would be nice to only read 1 (or some constant) number of DB entries during "tree update" and "generate proof" operations.

File: core-rust/state-manager/src/accumulator_tree/tree_builder.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (malexandru-rdx) [reply core-rust/state-manager/src/accumulator_tree/tree_builder.rs:82]: Fair enough :)

File: core-rust/state-manager/src/accumulator_tree/storage.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (malexandru-rdx) [reply core-rust/state-manager/src/accumulator_tree/storage.rs:160]: I think reference might be more suitable, both here and generally. You can always clone afterwards.

I guess I'm more concerned about the `get_proof(index)` implementation (which can be implemented without the `M`).

File: core-rust/state-manager/src/accumulator_tree/tree_builder.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (malexandru-rdx) [reply core-rust/state-manager/src/accumulator_tree/tree_builder.rs:170]: Oops, sorry I missed that `get`. Nope, definitely not worth it.

File: core-rust/state-manager/src/accumulator_tree/storage.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jakrawcz-rdx) [reply core-rust/state-manager/src/accumulator_tree/storage.rs:160]: > I think reference might be more suitable, both here and generally.

That's an interesting view - but if I agree with it, then it actually follows that absolutely ALL apis (i.e. interfaces defined using `trait`s) in rust should only ever operate on references. But that's not what I observe.
I try to gauge what is the convention on "when to use ref vs val", and so far it looks like it is driven more by "conceptual" reasons (also "ownership" reasons?) than by pure performance. I mean, e.g. DB access is supposed to "produce a value" (out of some external source/representation), so it returns a val. _(in fact, our in-memory impl could only benefit from `&` because it happens to store the items directly [because it is only a test util/helper], while the real storage will be DB-backed, and returning a ref from there could be tricky)_

File: core-rust/state-manager/src/accumulator_tree/tree_builder.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jakrawcz-rdx) [reply core-rust/state-manager/src/accumulator_tree/tree_builder.rs:170]: > Nope, definitely not worth it.

Hah, you are the deadliest enemy of `.clone()` :D
I get it and I respect it.
But also: at a price of (in practice) 32-byte memcpy, we could get rid of `zero: M` (to which many future readers will react "wait, why the guy has to store a zero here?!!".
And we don't even know if that `.clone()` costs anything in runtime (maybe the compiler notices "hey let's reuse this mem").
And we also kinda disregard that `zero: M` seen in this line requires a... 32-byte mem-cpy when we create an instance of `RelativeNodeAccess` :) (so we save a potential >=0 memcpy there at the price of guaranteed 1 memcpy here)

So to me, maybe it's worth it to return a val from `get()` after all?

[the remark this answers is not in the document]
Comment (malexandru-rdx) [reply core-rust/state-manager/src/accumulator_tree/tree_builder.rs:170]: > "wait, why the guy has to store a zero here?!!"
This indeed is confusing first time encountering it.

Probably this might be towards the right solution: https://docs.rs/generic_static/latest/generic_static/ ?

I'm more worried about the `slice_level: TreeSliceLevel<M>` than `zero` when creating a new instance. I imagine an "iterator"-like structure would not own a slice of data, but be used alongside the data structure itself.

[the remark this answers is not in the document]
Comment (jakrawcz-rdx) [reply core-rust/state-manager/src/accumulator_tree/tree_builder.rs:170]: I remember when I first saw `lazy_static`. I thought "wow, there are more people than just me who try to "fix" rust!". And surely there are :) Let me check that `generic_static` thing for `M::zero()` purposes. _(of course it does not come for free; I see there is a rwLock involved + a hashmap access - which makes me think it is mostly meant for caching expensive objects, not 32-bytes [of "fill 0"])_

> I imagine an "iterator"-like structure

I am genuinely curious to learn that pattern.

But my point was: should we be actively worried about [logical] "moves" of data when we create helper objects in rust? How do even know that compiler does not simply place things at the right mem offsets, so as to make the `TreeSliceLevel::new(...)` 0-cost?
_(I remember worrying about allocation [or rather GC] in certain java OO-heavy logic, until I saw that TLABs optimize all of it away - as if java had on-heap structs)_

I can feel your deep expertise on runtime performance. I totally see how many aspects of this PR could be micro-optimized.
But we cannot focus on everything at once. I think version 1 needs to focus on readability - so your comment on `zero: M,` being "weird" is very valid and I want to improve it. It is of course important to also avoid anything obviously sub-optimal (like doing O(n^2) when not necessary; or rewriting growing data volumes in storage). But optimizing anything more than that can be postponed - ideally after we identify bottlenecks.

[the remark this answers is not in the document]
Comment (jakrawcz-rdx) [reply core-rust/state-manager/src/accumulator_tree/tree_builder.rs:170]: There is no clear "best" way here, so let me summarize rough options and gather more opinions:

The problem:
A "logic helper" struct `RelativeNodeAccess<M>` has a `get M by index` method, which sometimes legitimately wants to return a meaningful "zero" object. Zero can be constructed statically (for a known `M` type). The callers of that `get` only want a reference.

How to do it?

A) Construct `M::zero()` on creation of `RelativeNodeAccess<M>` and store it as a [weird] part of that struct, in order to be able to return `&M` from `get()`. _(this option is currently seen in this PR)_

B) Return `M::zero()` directly from `get()` [even though no caller wants a value, they only want a ref]

C) Put `M::zero()` into a "generic static" structure [i.e. introduce dep https://docs.rs/generic_static/latest/generic_static/ and accept its runtime RwLock + HashMap implications], then easily return `&M` from `get()`.

@LukasGasior1 wdyt?

[the remark this answers is not in the document]
Comment (LukasGasior1) [reply core-rust/state-manager/src/accumulator_tree/tree_builder.rs:170]: I'd be leaning towards B), but happy to keep it as is too. Introducing generic_static seems a bit of an overkill here

[the remark this answers is not in the document]
Comment (jakrawcz-rdx) [reply core-rust/state-manager/src/accumulator_tree/tree_builder.rs]: Alright, thanks 👍 

Seems like A) is the good compromise option that everyone can live with.
I have added a comment on `zero: M` field in struct, so that it does not feel that weird in future.

[the remark this answers is not in the document]
Comment (malexandru-rdx) [reply core-rust/state-manager/src/accumulator_tree/tree_builder.rs]: > or rewriting growing data volumes in storage). But optimizing anything more than that can be postponed - ideally after we identify bottlenecks.
I agree, not worth optimizing the constant unless something outrageous (which is not the case here). My initial comment was missing the `get` and assumed that `zero: M` was needed just for the type transport (`PhantomData`) but I was wrong :).

I tend towards A) as well.

File: core-rust/state-manager/src/accumulator_tree/test.rs [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+fn simple_tree_produces_root_using_merkle_rule() {
+    let left = Hash([7; 32]);
+    let right = Hash([13; 32]);
+    let merkle_hash = blake2b_256_hash([left.0, right.0].concat());
+    let mut store = MemoryAccuTreeStore::new();
+    let root = append_at_index(&mut store, 0, vec![left, right]);
+    assert_eq!(root, merkle_hash);
+}
+
+// Detailed unit tests covering corner-cases, with a "fake hashing" allowing for easy inspection.
```
[the remark this answers is not in the document]
Comment (LukasGasior1) [reply core-rust/state-manager/src/accumulator_tree/test.rs:90]: :heart: love these tests :D so readable

**Review by LukasGasior1 — APPROVED** [review]

**Comment by LukasGasior1** [comment]
> I am not very convinced with the memory layout.

> I think it would shed more light/be closer to the final PR if proof generation would be taken into account from the start. With the current setup seems it will be kind of difficult to do it properly. Here we need to check out if we will only ever need proof for single a transaction or ranges as well (probably we're going to need it when doing ledger sync).

@malexandru-rdx 

I don't think we'll use these proofs for ledger sync. For both "regular" and "light" (in the future) sync the nodes should be able to build the tree themselves (and verify the root hash against a ledger proof)
